pkg/auth/providerrefresh/refresher.go

```diff
@@ -164,71 +159,56 @@ func (r *refresher) refreshAttributes(attribs *v3.UserAttribute) (*v3.UserAttrib
 
 	user, err := r.userLister.Get("", attribs.Name)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("error getting user %s: %w", attribs.Name, err)
 	}
 
-	loginTokens = make(map[string][]accessor.TokenAccessor)
-	derivedTokens = make(map[string][]accessor.TokenAccessor)
-
-	// List all tokens, v3 and ext.
-	// For ext tokens we actually filter for the user here.
+	loginTokens := make(map[string][]accessor.TokenAccessor)
+	derivedTokens := make(map[string][]accessor.TokenAccessor)
 
-	allV3Tokens, err := r.tokenLister.List("", labels.Everything())
-	if err != nil {
-		return nil, err
+	assign := func(token accessor.TokenAccessor) {
+		provider := token.GetAuthProvider()
+		if token.GetIsDerived() {
+			derivedTokens[provider] = append(derivedTokens[provider], token)
+			derivedTokenList = append(derivedTokenList, token)
+		} else {
+			loginTokens[provider] = append(loginTokens[provider], token)
+		}
 	}
 
-	allExtTokens, err := r.extTokenStore.ListForUser(user.Name)
+	// List v3.Tokens.
+	tokenUserIDLabelSet := labels.Set(map[string]string{tokens.UserIDLabel: user.Name})
+	v3Tokens, err := r.tokenLister.List("", tokenUserIDLabelSet.AsSelector())
```

Comment:
Note: There is a potential issue with this. There is however small chance that some one has non-expiring tokens that predate the label.